mathesar/views/frontend.py

```diff
@@ -1,31 +1,84 @@
 
 
-def get_schemas(request):
+def get_schema_list(request, database):
     schema_serializer = SchemaSerializer(
-        Schema.objects.all(),
+        Schema.objects.filter(database=database),
         many=True,
         context={'request': request}
     )
     return schema_serializer.data
 
 
-def index(request, **kwargs):
-    requested_db = kwargs.get('dbname')
-    database_list = get_non_default_database_keys()
-    if (requested_db is None or requested_db not in database_list) and len(database_list) > 0:
-        return redirect('index', dbname=database_list[0])
-    return render(
-        request,
-        "mathesar/index.html",
-        {
-            "common_data": {
-                "schemas": get_schemas(request),
-                "databases": database_list
-            }
-        }
+def get_database_list(request):
+    database_serializer = DatabaseSerializer(
+        Database.objects.all(),
+        many=True,
+        context={'request': request}
     )
+    return database_serializer.data
+
+
+def get_common_data(request, database, schema=None):
+    return {
+        'selected_db': database.name if database else None,
+        'selected_schema': schema.id if schema else None,
+        'schemas': get_schema_list(request, database),
+        'databases': get_database_list(request)
+    }
+
+
+def get_current_database(request, db_name):
+    # if there's a DB name passed in, try to retrieve the database, or return a 404 error.
+    if db_name is not None:
+        return get_object_or_404(Database, name=db_name)
+    else:
+        try:
+            # Try to get the first database available
+            return Database.objects.order_by('id').first()
+        except Database.DoesNotExist:
+            return None
+
+
+def get_current_schema(request, schema_id, database):
+    # if there's a schema ID passed in, try to retrieve the schema, or return a 404 error.
+    if schema_id is not None:
+        return get_object_or_404(Schema, id=schema_id)
+    else:
+        try:
+            # Try to get the first schema in the DB
+            return Schema.objects.filter(database=database).order_by('id').first()
+        except Schema.DoesNotExist:
+            return None
+
+
+def render_schema(request, database, schema):
+    # if there's no schema available, redirect to the schemas page.
+    if not schema:
+        return redirect('schemas', db_name=database.name)
+    else:
+        # We are redirecting so that the correct URL is passed to the frontend.
+        return redirect('schema_home', db_name=database.name, schema_id=schema.id)
+
+
+def home(request):
+    database = get_current_database(request, None)
+    schema = get_current_schema(request, None, database)
+    return render_schema(request, database, schema)
+
+
+def db_home(request, db_name):
+    database = get_current_database(request, db_name)
+    schema = get_current_schema(request, None, database)
+    return render_schema(request, database, schema)
```

Comment:
@kgodey If schemas are not synced yet while request is made, or if there are no databases configured, this will result in an infinite loop of redirection leading to `too many redirects` error and the browser killing the request.

We need to handle this for edge cases where schemas may not have synced.

I understand that no db is not an immediate concern, and we need atleast one db to start mathesar at the moment. In the future, if we decide allowing mathesar to start without a db and then configure databases using the frontend, this needs to be handled.

Reply:
Also all calls to `<db_name>/schemas/` would endup redirecting to the first available schema, but it should go to the schemas route, instead.

Reply:
@pavish I made `schemas` a separate view, which should fix both issues.

Reply:
@kgodey Looks good to me! Thank you.